File: n512.ml
type nat = { n : 'a. 'a -> ('a -> 'a) -> 'a } [@@ocaml.unboxed]

let zero () = { n = (fun z s -> z) }
let succ pred = { n = (fun z s -> s (pred.n z s)) }
let add a b = a.n b succ
let mul a b = a.n (zero ()) (fun n -> add n b)
let one () = succ (zero ())
let two () = succ (one ())
let n4 () = mul (two ()) (two ())
let n8 () = mul (n4 ()) (two ())
let n16 () = mul (n8 ()) (two ())
let n32 () = mul (n16 ()) (two ())
let n64 () = mul (n32 ()) (two ())
let n128 () = mul (n64 ()) (two ())
let n256 () = mul (n128 ()) (two ())
let n512 () = mul (n256 ()) (two ())
let n = n512 ()

File: hacking/n512.js
var zero = function () { return function (z) { return function (s) { return z; }; }; };
var succ = function (pred) {
    return function (z) {
        return function (s) {
            return s(pred(z)(s));
        };
    };
};
var add = function (a) { return function (b) { return a(b)(succ); }; };
var mul = function (a) { return function (b) { return a(zero())(function (n) { return add(n)(b); }); }; };
var one = function () { return succ(zero()); };
var two = function () { return succ(one()); };
var n4 = function () { return mul(two())(two()); };
var n8 = function () { return mul(n4())(two()); };
var n16 = function () { return mul(n8())(two()); };
var n32 = function () { return mul(n16())(two()); };
var n64 = function () { return mul(n32())(two()); };
var n128 = function () { return mul(n64())(two()); };
var n256 = function () { return mul(n128())(two()); };
var n512 = function () { return mul(n256())(two()); };
var bench = function (f) {
    var t1 = performance.now();
    f();
    console.log(performance.now() - t1);
};
console.log(bench(n512));
